hugr-model/src/v0/scope/vars.rs

```diff
@@ -0,0 +1,155 @@
+use fxhash::FxHasher;
+use indexmap::IndexSet;
+use std::hash::BuildHasherDefault;
+use thiserror::Error;
+
+use crate::v0::{NodeId, VarId};
+
+type FxIndexSet<K> = IndexSet<K, BuildHasherDefault<FxHasher>>;
+
+/// Table for keeping track of node parameters.
+///
+/// Variables refer to the parameters of a node which introduces a symbol.
+/// Variables have an associated name and are scoped via nodes. The types of
+/// parameters of a node may only refer to earlier parameters in the same node
+/// in the order they are defined. A variable name must be unique within a
+/// single node. Each node that introduces a symbol introduces a new isolated
+/// scope for variables.
+///
+/// # Examples
+///
+/// ```
+/// # pub use hugr_model::v0::{NodeId, VarId};
+/// # pub use hugr_model::v0::scope::VarTable;
+/// let mut vars = VarTable::new();
+/// vars.enter(NodeId(0));
+/// vars.insert("foo").unwrap();
+/// assert_eq!(vars.resolve("foo").unwrap(), VarId(NodeId(0), 0));
+/// assert!(!vars.is_visible(VarId(NodeId(0), 1)));
+/// vars.insert("bar").unwrap();
+/// assert!(vars.is_visible(VarId(NodeId(0), 1)));
+/// assert_eq!(vars.resolve("bar").unwrap(), VarId(NodeId(0), 1));
+/// vars.enter(NodeId(1));
+/// assert!(vars.resolve("foo").is_err());
+/// assert!(!vars.is_visible(VarId(NodeId(0), 0)));
+/// vars.exit();
+/// assert_eq!(vars.resolve("foo").unwrap(), VarId(NodeId(0), 0));
+/// assert!(vars.is_visible(VarId(NodeId(0), 0)));
+/// ```
+#[derive(Debug, Clone)]
+pub struct VarTable<'a> {
+    vars: FxIndexSet<(NodeId, &'a str)>,
+    scopes: Vec<VarScope>,
+}
+
+impl<'a> VarTable<'a> {
+    /// Create a new empty variable table.
+    pub fn new() -> Self {
+        Self {
+            vars: FxIndexSet::default(),
+            scopes: Vec::new(),
+        }
+    }
+
+    /// Enter a new scope for the given node.
+    pub fn enter(&mut self, node: NodeId) {
+        self.scopes.push(VarScope {
+            node,
+            var_count: 0,
+            var_stack: self.vars.len(),
+        })
+    }
+
+    /// Exit a previously entered scope.
+    ///
+    /// # Panics
+    ///
+    /// Panics if there are no open scopes.
+    pub fn exit(&mut self) {
+        let scope = self.scopes.pop().unwrap();
+        self.vars.drain(scope.var_stack..);
+    }
+
+    /// Resolve a variable name to a node and variable index.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the variable is not defined in the current scope.
+    ///
+    /// # Panics
+    ///
+    /// Panics if there are no open scopes.
+    pub fn resolve(&self, name: &'a str) -> Result<VarId, UnknownVarError<'a>> {
+        let scope = self.scopes.last().unwrap();
+        let (set_index, _) = self
+            .vars
+            .get_full(&(scope.node, name))
```

Comment:
Does `get_full` return the pair? Can you do `.get` and then not have to discard the second element (as in `let (set_index, _) = ...`)?

Reply:
This code was written before I knew of `get_index_of`. I replaced it with `get_index_of` now.